src/server.rs

```diff
@@ -559,15 +575,29 @@ fn collect_removals(
     removal_buffer: &RemovalBuffer,
 ) -> Result<()> {
     for (&entity, remove_ids) in removal_buffer.iter() {
-        let entity_range = serialized.write_entity(entity)?;
-        let ids_len = remove_ids.len();
-        let fn_ids = serialized.write_fn_ids(remove_ids.iter().map(|&(_, fns_id)| fns_id))?;
-        for (client_entity, mut message, .., visibility) in &mut *clients {
-            if !visibility.is_hidden(entity) {
-                trace!(
-                    "writing removals for `{entity}` with `{remove_ids:?}` for client `{client_entity}`"
-                );
-                message.add_removals(entity_range.clone(), ids_len, fn_ids.clone());
+        let mut entity_range = None;
+        for (_, mut message, ..) in &mut *clients {
+            message.start_entity_removals();
+        }
+
+        for &(component_id, fns_id) in remove_ids {
+            let mut fns_id_range = None;
+            for (client_entity, mut message, .., mut ticks, _, _) in &mut *clients {
+                // Only send removals for components that were previously sent.
+                let Some(entity_ticks) = ticks.entities.get_mut(&entity) else {
+                    continue;
+                };
+                if !entity_ticks.components.remove(&component_id) {
+                    continue;
+                }
```

Comment:
Should also filter on `visibility.hidden(entity)` to avoid redundantly sending component removals when visibility is lost (which will also send a despawn).

Reply:
Added a comment about it.